### PR TITLE
Backend question validation on Creation and Edit

### PR DESCRIPTION
**Pull request by mehtadarshan1**
Fixes issue #107

### CODE REVIEW

File: server/questions.js [hunk 1]
```diff
@@ -64,13 +76,13 @@ exports.addQuestionByType = function(qType, question, callback) {
...
 
-	// validate question by its type
-	var result = validateQuestionByType(questionToAdd, questionToAdd.type);
-	if (result !== null){
-		return callback({status:400, msg:result}, null)
+	// validate constant question attributes
+	result = validator.questionCreationValidation(questionToAdd);
+	if (result.success){
+		db.addQuestion(questionToAdd, callback);
+	} else{
```
Comment (esteeroramy) [inline server/questions.js]: consistent spacing

**Review by esteeroramy — APPROVED** [review]

**Review by Mohammed-JS-Khan — APPROVED** [review]
Looks gucci 👌

File: server/common.js [hunk 2]
```diff
@@ -41,4 +41,23 @@ const userTypes = Object.freeze({
     ADMIN     : 0,
     STUDENT   : 1
 });
-exports.userTypes = userTypes;
\ No newline at end of file
+
+exports.userTypes = userTypes;
+
+const questionAttributes = Object.freeze({
```
Comment (ahmedqarmout2) [inline server/common.js]: good point. lets put them as common, but we also need the other question types' attributes here. lets rename this to general question attributes, then we can add regular, multiple choice true and false etc... as their own question attributes

File: server/validation.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
+/*
+question.js
```
Comment (ahmedqarmout2) [inline server/validation.js]: rename to questionValidator (the file as well). we are going to need a user validator later on.

File: server/questions.js [hunk 4]
```diff
@@ -21,7 +21,17 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 var db = require('./db.js');
 var logger = require('./log.js').logger;
 var common = require('./common.js');
+var validator = require('./validation.js');
```
Comment (ahmedqarmout2) [inline server/questions.js]: rename to questionValidator

File: server/questions.js [hunk 5]
```diff
@@ -21,7 +21,17 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
 
+var prepareQuestionData = function(data){
```
Comment (ahmedqarmout2) [inline server/questions.js]: this is a start of a great idea, I love where this is going. Lets have prepareQuestionData(data, callback) In this function we will do the validation and if t he validation passes we can return a pure object of what the question needs to be created/edited. For this we will need different prepare function for each question type. So like, prepareMultipleChoiceQuestionData(data, callback) and so on.

File: server/validation.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+var common = require('./common.js');
+
+var validateAttributeType = function(valueToCheck,key){
+	if(Object.prototype.toString.call(valueToCheck) !== common.questionAttributes[key].type)
```
Comment (ahmedqarmout2) [inline server/validation.js]: return Object.....===common.... ;

File: server/questions.js [hunk 7]
```diff
@@ -36,15 +46,17 @@ exports.addQuestionByType = function(qType, question, callback) {
 	questionToAdd.text = question.text;
 	questionToAdd.answer = question.answer;
 	questionToAdd.hint = question.hint;
-	questionToAdd.points = parseInt(question.points);
-	questionToAdd.visible = (question.visible === 'true');
+	questionToAdd.points = question.points
```
Comment (ahmedqarmout2) [inline server/questions.js]: forgot ;

**Review by ahmedqarmout2 — CHANGES_REQUESTED** [review]
Great work. I like where this is going, but we need to make a lot of changes

File: server/questions.js (hunk 5 above)
Comment (mehtadarshan1) [reply server/questions.js]: That is a good point, what we can do here is, where I have created functions for each question type to validate (i.e. multipleChoiceValidator,trueAndFalseValidator) We can have it prepare the data and validate it as well. This saves extra code from being created.

File: server/common.js (hunk 2 above)
Comment (mehtadarshan1) [reply server/common.js]: That is right, we should have other data types in this as well under different key/values.

File: server/db.js [hunk 8]
```diff
@@ -594,8 +594,8 @@ exports.updateQuestionById = function(questionId, request, callback){
         }
     }
     
-    if (request.visible) {
-        update.$set.visible = (request.visible === 'true');
+    if ('visible' in request) {
```
Comment (ahmedqarmout2) [inline server/db.js]: I wonder if we can do this for the rest of the request params

File: server/questions.js [hunk 9]
```diff
@@ -36,13 +36,13 @@ var questionUpdateParser = function(question){
 }
 var prepareQuestionData = function(question, callback){
 	// prepare regular data
+	console.log(question);
```
Comment (esteeroramy) [inline server/questions.js]: Should remove console.log

Comment (mehtadarshan1) [reply server/questions.js]: good one

File: server.js [hunk 10]
```diff
@@ -712,7 +712,7 @@ app.put('/questionadd', function(req, res) {
         return res.redirect('/');
     }
 
-    questions.addQuestionByType(req.body.type, req.body, function(err, qId) {
+    questions.addQuestionByType(req.body, function(err, qId) {
```
Comment (ahmedqarmout2) [inline server.js]: addQuestionByType >> addQuestion

File: server/questionValidator.js [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+const failMsg = {success:false, msg:'Fields types are Incorrect'}
+
+var validateAttributeType = function(valueToCheck, key, attributeType){
+	return Object.prototype.toString.call(valueToCheck) === common.questionAttributes[attributeType][key].type;
+}
+
+/*Validate all question fields on first entry to db*/
+exports.questionCreationValidation = function(info) {
+	for (var key in common.questionAttributes.DEFAULT){
+		if (!(key in info) || !validateAttributeType(info[key], key, 'DEFAULT'))
```
Comment (ahmedqarmout2) [inline server/questionValidator.js]: if (.....) {
}

File: server/questionValidator.js [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+			return failMsg
```
Comment (ahmedqarmout2) [inline server/questionValidator.js]: ;

File: server/questionValidator.js [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+/*Validate all fields that will be modified*/
+exports.validateAttributeFields = function(question,type){
+	var extraAttributes = false
+
+	for (var key in question){
+		// check const attributes
+		if (key in common.questionAttributes.DEFAULT){
+			if (!validateAttributeType(question[key], key, 'DEFAULT')){
+				return failMsg
```
Comment (ahmedqarmout2) [inline server/questionValidator.js]: ;

File: server/questionValidator.js [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+			}
+		// not common field found
+		} else {
+			extraAttributes = true
```
Comment (ahmedqarmout2) [inline server/questionValidator.js]: ;

File: server/questionValidator.js [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+		}
+	}
+
+	// check by question type to validate the extra fields
+	if (extraAttributes){
+		return validateQuestionAttributesByType(question,type);
+	}
+	return successMsg
```
Comment (ahmedqarmout2) [inline server/questionValidator.js]: ;

File: server/questionValidator.js [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+var regexAttributeValidator = function(question){
+	if (!question.answer || !validateAttributeType(question.answer,'answer','REGULAR')){
+		return qTypeFailMsg('Missing answer field.');
+	}
+	return successMsg;
+}
+
+var multipleChoiceAttributeValidator = function(question){
+	if (!question.choices || question.choices.length < 2){
```
Comment (ahmedqarmout2) [inline server/questionValidator.js]: validate each choice is a string

File: server/questionValidator.js [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+
+var trueAndFalseAttributeValidator = function(question){
+	if (question.choices && question.choices.length !== 2){
+		return qTypeFailMsg('True and False can only have 2 options!');
+	}
+	return successMsg;
+}
+
+var matchingAttributeValidator = function(question){
+	// need to check for left and right columns
```
Comment (ahmedqarmout2) [inline server/questionValidator.js]: reminder to implement this

File: server/questionValidator.js [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+
+var multipleChoiceAttributeValidator = function(question){
+	if (!question.choices || question.choices.length < 2){
+		return qTypeFailMsg('Need two or more options for Multiple Choice Question!');
+	}
+	return successMsg;
+}
+
+var trueAndFalseAttributeValidator = function(question){
+	if (question.choices && question.choices.length !== 2){
```
Comment (ahmedqarmout2) [inline server/questionValidator.js]: if (!question.choices || question.choices.length !== 2){
...
}

File: server/questions.js [hunk 19]
```diff
@@ -21,20 +21,28 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+	}
+	if ('points' in question){
+		updatedQuestion.points = parseInt(question.points);
+	}
+	
+	return updatedQuestion;
+}
+var prepareQuestionData = function(question, callback){
+	// prepare regular data
+	console.log(question);
```
Comment (ahmedqarmout2) [inline server/questions.js]: remove log statement

File: server/questions.js [hunk 20]
```diff
@@ -45,57 +53,59 @@ exports.addQuestionByType = function(qType, question, callback) {
...
-* 3rd bit: sort by points
-*
-* QUERYING
-* 4th bit: if 1, allow questions that have already been answered by user.
-* 5th bit: if 1, only show those questions that have been answered (with bit 4).
+* Insert a new regular question into the database.
+* The question object passed to the function should have
+* the text, topic, type, answer, points and hint set.
 */
+exports.addQuestionByType = function(question, callback) {
```
Comment (ahmedqarmout2) [inline server/questions.js]: addQuestionByType >> addQuestion

File: server/questions.js [hunk 21]
```diff
@@ -206,26 +183,26 @@ exports.checkAnswer = function(questionId, user, answer, callback) {
...
-					var checkIndexRight;
-
-					for (i = 0; i < ansLeftSide.length; i++) {
-					    checkIndexLeft = question.leftSide.indexOf(ansLeftSide[i]);
-							checkIndexRight = question.rightSide.indexOf(ansRightSide[i]);
-							if (checkIndexLeft !== checkIndexRight) {
-							    value = false;
-							}
-					}
+			const ansRightSide = answer[1];
```
Comment (ahmedqarmout2) [inline server/questions.js:184]: hmmm this looks a little off, could you double check that its all aligned

Comment (mehtadarshan1) [reply server/questions.js:184]: I have aligned it properly, github changes the formatting

File: server/questions.js [hunk 22]
```diff
@@ -121,50 +131,17 @@ var updateQuestionByType = function(qId, infoToUpdate, callback){
 		if(err){
```
Comment (ahmedqarmout2) [inline server/questions.js:129]: change this function to updateQuestionById instead of type

File: server/questions.js [hunk 23]
```diff
@@ -206,26 +183,26 @@ exports.checkAnswer = function(questionId, user, answer, callback) {
...
-					if (value === null) {
-						  value = true;
-					}
-				} else {
-					value = false
+				if (value === null) {
+					  value = true;
 				}
+			} else {
+				value = false
```
Comment (ahmedqarmout2) [inline server/questions.js]: ;

File: server/questions.js [hunk 24]
```diff
@@ -206,26 +183,26 @@ exports.checkAnswer = function(questionId, user, answer, callback) {
...
-							}
-					}
+			const ansRightSide = answer[1];
+
+			if (ansLeftSide.length === question.leftSide.length) {
+				var checkIndexLeft;
+				var checkIndexRight;
+
+				for (i = 0; i < ansLeftSide.length; i++) {
+				    checkIndexLeft = question.leftSide.indexOf(ansLeftSide[i]);
```
Comment (ahmedqarmout2) [inline server/questions.js:191]: this seems off :(

File: server/questions.js [hunk 25]
```diff
@@ -21,20 +21,28 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
 var common = require('./common.js');
+var questionValidator = require('./questionValidator.js');
 
-/*
-* Insert a new regular question into the database.
-* The question object passed to the function should have
-* the text, topic, type, answer, points and hint set.
-*/
-exports.addQuestionByType = function(qType, question, callback) {
+var questionUpdateParser = function(question){
```
Comment (ahmedqarmout2) [inline server/questions.js]: comments on functions plz

File: server/questions.js (hunk 24 above)
Comment (mehtadarshan1) [reply server/questions.js:191]: GitHub changes the view of the format

File: server/questions.js (hunk 22 above)
Comment (mehtadarshan1) [reply server/questions.js:129]: done

File: server/common.js [hunk 26]
```diff
@@ -44,6 +44,51 @@ const userTypes = Object.freeze({
...
+
+    },
+    MATCHING: {
+        leftSide    : {type:'[object Array]'},
+        rightSide   : {type:'[object Array]'}
+    },
+    DATATYPES: {
+        Array       : {type:'[object Array]'},
+        String      : {type:'[object String]'},
+        Number      : {type:'[object Number]'},
```
Comment (ahmedqarmout2) [inline server/common.js:85]: Should we split this into integers, floats etc..?

**Review by ahmedqarmout2 — APPROVED** [review]
Just one last question

**Review by Mohamed-A-Mohamed — APPROVED** [review]
Nice work

**Pull request merged**